### PR TITLE
Don't polyfill `process` on Middleware

### PR DESCRIPTION
**Pull request by Schniz**
since Middleware run with a different runtime, the polyfills work like their browser counterparts,
which make no sense in a server-side environment and only get in the way.

This commit introduces a warning when using anything from `process` other than `process.env`,
and returning `undefined` so users won't use the polyfilled `process`

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 65244665ca761eaf5d97f84bbbae15de16186981

`yarn testheadless test/integration/react-streaming-and-server-components/test/switchable-runtime.test.js` 
- Switchable runtime (prod) > should build /edge as a dynamic page with the edge runtime
- Switchable runtime (prod) > should build /edge-rsc as a dynamic page with the edge runtime

<details>
<summary>Expand output</summary>

  ● Switchable runtime (prod) › should build /edge as a dynamic page with the edge runtime

    TypeError: Cannot read property '1' of null

      25 | async function testRoute(appPort, url, { isStatic, isEdge, isRSC }) {
      26 |   const html1 = await renderViaHTTP(appPort, url)
    > 27 |   const renderedAt1 = +html1.match(/Time: (\d+)/)[1]
         |                                                  ^
      28 |   expect(html1).toContain(`Runtime: ${isEdge ? 'Edge' : 'Node.js'}`)
      29 |
      30 |   const html2 = await renderViaHTTP(appPort, url)

      at testRoute (integration/react-streaming-and-server-components/test/switchable-runtime.test.js:27:50)
      at Object.<anonymous> (integration/react-streaming-and-server-components/test/switchable-runtime.test.js:150:5)

  ● Switchable runtime (prod) › should build /edge-rsc as a dynamic page with the edge runtime

    TypeError: Cannot read property '1' of null

      25 | async function testRoute(appPort, url, { isStatic, isEdge, isRSC }) {
      26 |   const html1 = await renderViaHTTP(appPort, url)
    > 27 |   const renderedAt1 = +html1.match(/Time: (\d+)/)[1]
         |                                                  ^
      28 |   expect(html1).toContain(`Runtime: ${isEdge ? 'Edge' : 'Node.js'}`)
      29 |
      30 |   const html2 = await renderViaHTTP(appPort, url)

      at testRoute (integration/react-streaming-and-server-components/test/switchable-runtime.test.js:27:50)
      at Object.<anonymous> (integration/react-streaming-and-server-components/test/switchable-runtime.test.js:158:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| buildDuration | 16s | 16.2s | ⚠️ +232ms |
| buildDurationCached | 6.2s | 6.2s | ⚠️ +55ms |
| nodeModulesSize | 481 MB | 481 MB | ⚠️ +1.79 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.283 | 3.309 | ⚠️ +0.03 |
| / avg req/sec | 761.53 | 755.45 | ⚠️ -6.08 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.414 | 1.429 | ⚠️ +0.02 |
| /error-in-render avg req/sec | 1768.52 | 1749.25 | ⚠️ -19.27 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.3 kB | 72.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.04 kB | 3.04 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.73 kB | 5.73 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| _buildManifest.js gzip | 461 B | 461 B | ✓ |
| Overall change | 461 B | 461 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| buildDuration | 18.5s | 18.6s | ⚠️ +72ms |
| buildDurationCached | 6.2s | 6.3s | ⚠️ +88ms |
| nodeModulesSize | 481 MB | 481 MB | ⚠️ +1.79 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.3 | 3.303 | 0 |
| / avg req/sec | 757.57 | 757 | ⚠️ -0.57 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.404 | 1.399 | 0 |
| /error-in-render avg req/sec | 1780 | 1787 | +7 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.1 kB | 29.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.9 kB | 72.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.02 kB | 3.02 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.78 kB | 5.78 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.44 kB | 2.44 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 393 B | 393 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| _buildManifest.js gzip | 457 B | 457 B | ✓ |
| Overall change | 457 B | 457 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | Schniz/next.js warn-for-nodejs-runtime-apis-in-middleware | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 65244665ca761eaf5d97f84bbbae15de16186981

File: test/production/using-node.js-runtime-apis-fails-to-build/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import { nextBuild } from 'next-test-utils'
```
Comment (Schniz) [inline test/production/using-node.js-runtime-apis-fails-to-build/index.test.ts:1]: add `in-edge` to filename

**Comment by Schniz** [comment]
superseded by #36980

**Pull request closed without merging**